packages/connection/index.js

```diff
@@ -106,12 +107,25 @@ class Connection extends EventEmitter {
     return this.jid
   }
 
-  _online() {
-    this.emit('online', this.jid)
+  _status(status, ...args) {
+    this.status = status
+    this.emit('status', status, ...args)
+    this.emit(status, ...args)
   }
 
-  _authenticated() {
-    this.emit('authenticated')
+  ready(fn) {
+    if (fn) {
+      if (this.status === 'online') {
+        return Promise.resolve().then(fn)
+      }
+      this.on('online', fn)
+      return
+    }
+
+    if (this.status === 'online') {
+      return Promise.resolve()
+    }
+    return this.once('online')
```

Comment:
Observable could be used as well for the use case of running code if true and when true. 

example
`entity.subscribe('ready', fn)` returns a subscription